prelude.dx

```diff
@@ -581,3 +588,83 @@ def (<>) (x:List a) (y:List a) : List a =
     case i' < nx of
       True  -> xs.(fromOrdinal _ i')
       False -> ys.(fromOrdinal _ (i' - nx))
+
+'Isomorphisms
+
+data Iso a:Type b:Type = MkIso { fwd: a -> b & bwd: b -> a }
+
+def appIso (iso: Iso a b) (x:a) : b =
+  (MkIso {fwd, bwd}) = iso
+  fwd x
+
+def flipIso (iso: Iso a b) : Iso b a =
+  (MkIso {fwd, bwd}) = iso
+  MkIso {fwd=bwd, bwd=fwd}
+
+def revIso (iso: Iso a b) (x:b) : a = appIso (flipIso iso) x
+
+idIso : Iso a a = MkIso {fwd=id, bwd=id}
+
+def (<>>) (iso1: Iso a b) (iso2: Iso b c) : Iso a c =
+  (MkIso {fwd=fwd1, bwd=bwd1}) = iso1
+  (MkIso {fwd=fwd2, bwd=bwd2}) = iso2
+  MkIso {fwd=(fwd1 >>> fwd2), bwd=(bwd1 <<< bwd2)}
+
+def (<<>) (iso2: Iso b c) (iso1: Iso a b) : Iso a c = iso1 <>> iso2
+
+-- Lens-like accessors
+-- (note: #foo is an Iso {foo: a & ...b} (a & {&...b}))
+def getAt  (iso: Iso a (b & c)) : a -> b = fst <<< appIso iso
+def popAt  (iso: Iso a (b & c)) : a -> c = snd <<< appIso iso
+def pushAt (iso: Iso a (b & c)) (x:b) (r:c) : a = revIso iso (x, r)
+def setAt  (iso: Iso a (b & c)) (x:b) (r:a) : a =
+  pushAt iso x $ popAt iso r
+
+-- Prism-like accessors
+-- (note: #?foo is an Iso {foo: a | ...b} (a | {|...b}))
+def matchWith (iso: Iso a (b | c)) (x: a) : Maybe b =
+  case appIso iso x of
+    Left x -> Just x
+    Right _ -> Nothing
+def buildWith (iso: Iso a (b | c)) (x: b) : a = revIso iso $ Left x
+
+swapPairIso : Iso (a & b) (b & a) =
+  MkIso {fwd = \(a, b). (b, a), bwd = \(b, a). (a, b)}
+
+-- Complement the focus of a lens-like isomorphism
+exceptLens : Iso a (b & c) -> Iso a (c & b) = \iso. iso <>> swapPairIso
+
+swapEitherIso : Iso (a | b) (b | a) =
+  fwd = \x. case x of
+    Left l -> Right l
+    Right r -> Left r
+  bwd = \x. case x of
+    Left r -> Right r
+    Right l -> Left l
+  MkIso {fwd, bwd}
+
+-- Complement the focus of a prism-like isomorphism
+exceptPrism : Iso a (b | c) -> Iso a (c | b) = \iso. iso <>> swapEitherIso
+
+-- Use a lens-like iso to split a 1d table into a 2d table
+def over (iso: Iso a (b & c)) (tab: a=>v) : (b=>c=>v) =
+  for i j. tab.(revIso iso (i, j))
```

Comment:
That's really nice, although I'm not sure if the name `over` is that clear?

Reply:
Is `overLens` better? `lensToAxis`? `lensToLeading`?

Reply:
Changed to `overLens` for consistency with `overFields`